README
HTML-Tree version {{$version}}, released {{$date}}
{{ $zilla->is_trial
   ? "THIS IS A DEVELOPMENT RELEASE FOR TESTING PURPOSES ONLY.\n" : '' }}

This distribution contains a suite of modules for representing,
creating, and extracting information from HTML syntax trees; there is
also relevent documentation.  These modules used to be part of the
libwww-perl distribution, but are now unbundled in order to facilitate
a separate development track.  Bug reports and discussions about these
modules can be sent to the RT queue at <bug-html-tree@rt.cpan.org>.

The modules and documentation present in this collection are:

  HTML::Tree -- overview of the modules.

  HTML::Tree::AboutObjects -- an article by Sean M. Burke from
    The Perl Journal #17: "User's View of Object-Oriented Modules"

  HTML::Tree::AboutTrees -- an article by Sean M. Burke from
    The Perl Journal #18: "Trees", about tree-shaped data
    structures in Perl.

  HTML::Tree::Scanning -- an article by Sean M. Burke from
    The Perl Journal #19: "Scanning HTML".


  HTML::Element - class for representing the nodes of the HTML syntax
        trees.  The elements have other elements and text segments
        as children.  The HTML::Element class has methods to build,
        alter, and traverse the structure of the tree.

  HTML::TreeBuilder - a class that uses HTML::Parser to read HTML
        document text and build from it a syntax tree made of
        HTML::Element nodes.

  HTML::Parse - deprecated module.  Now just a wrapper around
        HTML::TreeBuilder

  HTML::AsSubs - module providing an easy way to build an HTML syntax
        tree by nesting functions.



INSTALLATION

To install this module, run the following commands:

{{ $t->build_instructions }}



DEPENDENCIES

{{ $t->dependency_list }}



SOURCE

The source is now kept at {{ $meta{resources}{repository}{web}
                             // die 'no repo' }}



SUBLIMINAL MESSAGE: BUY THE BOOK!

The book /Perl & LWP/ by Sean M. Burke is published by
O'Reilly and Associates, 2002.  ISBN: 0-596-00178-9

It has several chapters to do with HTML processing in general,
and HTML-Tree specifically.  There's more info at:

  http://www.oreilly.com/catalog/perllwp/
  http://www.amazon.com/exec/obidos/ASIN/0596001789



COPYRIGHT

Copyright 1995-1998 Gisle Aas; 1999-2004 Sean M. Burke;
2005 Andy Lester; 2006 Pete Krawczyk; 2010 Jeff Fearn;
2012 Christopher J. Madsen (Except the articles contained in
HTML::Tree::AboutObjects, HTML::Tree::AboutTrees, and
HTML::Tree::Scanning, which are all copyright 2000 The Perl Journal.)

Except for those three TPJ articles, the whole HTML-Tree distribution,
of which this file is a part, is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

Those three TPJ articles may be distributed under the same terms as
Perl itself.

The programs in this library are distributed in the hope that they
will be useful, but without any warranty; without even the implied
warranty of merchantability or fitness for a particular purpose.


Changes
Changelog for HTML-Tree

{{$NEXT}}

5.909    2014-05-24  Trial Release by Christopher J. Madsen

    [TESTS]
    - Skip LWP::UserAgent tests if the version installed is too old
    - Change 00system.t to use diag so the output is visible

5.908    2014-05-17  Trial Release by Christopher J. Madsen

    [ENHANCEMENTS]
    - New as_lol & content_as_lol methods in HTML::Element
      (Ben Deutsch, GitHub #2)

    [FIXES]
    - as_XML inserted extra whitespace in comments (Tomaž Solc, RT #94311)

5.907    2013-09-13  Trial Release by Christopher J. Madsen

    [ENHANCEMENTS]
    - New content_as_HTML & content_as_XML methods in HTML::Element

5.906    2013-07-13  Trial Release by Christopher J. Madsen

    [DOCUMENTATION]
    - Explain that implicit_tags => 0 still returns implicit <html> (RT #86062)
    - typo fixes supplied by David Steinbrunner (GitHub #1)

    [TESTS]
    - Add implicit-tags-off.t

5.905    2013-06-29  Trial Release by Christopher J. Madsen

    [FIXES]
    - as_XML would not allow single-character attribute names (RT #86566)

5.904    2013-06-08  Trial Release by Christopher J. Madsen

    [FIXES]
    - as_text would infinite loop if it encountered undef (RT #83757)

5.903    2013-06-01  Trial Release by Christopher J. Madsen

    [ENHANCEMENTS]
    - $HTML::Element::default_encoding gets its default value from
      $ENV{PERL_HTML_TREE_ENCODING}, so you can restore the old behavior
      without modifying any code (just export PERL_HTML_TREE_ENCODING='').

5.902    2013-03-01  Trial Release by Christopher J. Madsen

    [DOCUMENTATION]
    - minor clarifications

    [TESTS]
    - Add dump.t

5.901    2012-12-22  Trial Release by Christopher J. Madsen

    [TESTS]
    - Fix encoding.t to work on Perl 5.8.9 and 5.10.0

5.900    2012-12-15  Trial Release by Christopher J. Madsen

    [THINGS THAT MAY BREAK YOUR CODE OR TESTS]
    - parse_file (and new_from_file) now try to determine the encoding
      automatically when given a filename (not a filehandle).  To restore
      the old behavior, set the new encoding attribute to the empty string.
      To restore it globally, set $HTML::Element::default_encoding = ''.

    [ENHANCEMENTS]
    - new_from_file & new_from_url now let you set parsing attributes
    - New shortcut constructor new_from_string is like new_from_content,
      but allows you to set parsing attributes
    - New shortcut constructor new_from_http for constructing a tree
      from the content of a HTTP::Message (or subclass like HTTP::Response)
    - Setting the new self_closed_tags attribute to 1 makes TreeBuilder
      handle XML-style self-closed tags (e.g. <a id="a1" />)
    - New child_nodes method makes for simpler recursion
    - New openw and encode_fh methods for writing a file with the
      correct encoding

    [DOCUMENTATION]
    - new actually does take optional attributes
      (It has since at least 3.18, although undocumented, and it did
       not previously work with ignore_ignorable_whitespace.)
    - methods & attributes added in version 4.0 or later are now marked
    - don't recommend the traverse method; give recursive example (RT #48344)

    [TESTS]
    - Add test for self_closed_tags attribute.
    - Clarify skip message in construct_tree.t (RT #81371)


5.03     2012-09-22  Release by Christopher J. Madsen

    [THINGS THAT MAY BREAK YOUR CODE OR TESTS]
    - as_HTML no longer indents <textarea> (Tomohiro Hosaka) (RT #70385)

    [FIXES]
    - as_trimmed_text did not accept '0' for extra_chars

    [DOCUMENTATION]
    - Explain that as_text never adds whitespace (RT #66498)
    - Explain what extra_chars can contain for as_trimmed_text.


5.02     2012-06-27  Release by Christopher J. Madsen

    [TESTS]
    - Do not attempt to check result of $! in construct_tree.t
      (The fix in 5.01 was not successful.)


5.01     2012-06-20  Release by Christopher J. Madsen

    [TESTS]
    - Force C locale in construct_tree.t (in non-English locales,
      $! will produce messages in a different language) (RT #77823)
    - Add test for preserving whitespace while parsing.


5.00     2012-06-12  Release by Christopher J. Madsen

   There are only some minor documentation changes since 4.903.
   This is a summary of the most significant changes since 4.2.

    [THINGS THAT MAY BREAK YOUR CODE OR TESTS]
    - Use weak references to avoid memory leaks
      See "Weak References" in HTML::Element for details.
    - new_from_file now dies if the file cannot be opened.  $! records
      the specific problem.  (Previously, you got a tree with a few
      implicit elements.)
    - Some methods normally returning a scalar could return the empty
      list in certain circumstances.  This has been corrected.  The
      affected methods are: address, deobjectify_text, detach, is_inside,
      & pindex.
    - deprecate the Version sub/method.  Use the VERSION method instead.

    [ENHANCEMENTS]
    - add new_from_url (Using LWP) (David Pottage) (RT #68097)

    [DOCUMENTATION]
    - Explain that parse_file (and new_from_file) opens files in binary mode
      BUT THIS IS PLANNED TO CHANGE.


4.903    2012-06-08  Trial Release by Christopher J. Madsen

    [DOCUMENTATION]
    - Explain that parse_file (and new_from_file) opens files in binary mode
      BUT THIS IS PLANNED TO CHANGE.

    [TESTS]
    - test error handling for new_from_file & new_from_url
    - remove use_ok from most tests
      (if the module won't load, the tests can't pass anyway)


4.902    2012-06-06  Trial Release by Christopher J. Madsen

    [THINGS THAT MAY BREAK YOUR CODE OR TESTS]
    - new_from_url now dies if the request fails or the response is not HTML


4.901    2012-06-06  Trial Release by Christopher J. Madsen

    [THINGS THAT MAY BREAK YOUR CODE OR TESTS]
    - new_from_file now dies if the file cannot be opened.  $! records
      the specific problem.  (Previously, you got a tree with a few
      implicit elements.)

lib/HTML/Tree.pm
package HTML::Tree;

# ABSTRACT: build and scan parse-trees of HTML

# HTML::Tree is basically just a happy alias to HTML::TreeBuilder.

use warnings;
use strict;

# VERSION from OurPkgVersion

use HTML::TreeBuilder ();

=head1 SYNOPSIS

    use HTML::TreeBuilder;
    my $tree = HTML::TreeBuilder->new();
    $tree->parse_file($filename);

        # Then do something with the tree, using HTML::Element
        # methods -- for example:

    $tree->dump

        # Finally:

    $tree->delete;

=cut

sub new {
    shift;
    unshift @_, 'HTML::TreeBuilder';
    goto &HTML::TreeBuilder::new;
}

sub new_from_file {
    shift;
    unshift @_, 'HTML::TreeBuilder';
    goto &HTML::TreeBuilder::new_from_file;
}

sub new_from_content {
    shift;
    unshift @_, 'HTML::TreeBuilder';
    goto &HTML::TreeBuilder::new_from_content;
}

sub new_from_string {
    shift;
    unshift @_, 'HTML::TreeBuilder';
    goto &HTML::TreeBuilder::new_from_string;
}

sub new_from_url {
    shift;
    unshift @_, 'HTML::TreeBuilder';
    goto &HTML::TreeBuilder::new_from_url;
}

sub new_from_http {
    shift;
    unshift @_, 'HTML::TreeBuilder';
    goto &HTML::TreeBuilder::new_from_http;
}

1;

=head1 DESCRIPTION

HTML-Tree is a suite of Perl modules for making parse trees out of
HTML source.  It consists of mainly two modules, whose documentation
you should refer to: L<HTML::TreeBuilder|HTML::TreeBuilder>
and L<HTML::Element|HTML::Element>.

HTML::TreeBuilder is the module that builds the parse trees.  (It uses
HTML::Parser to do the work of breaking the HTML up into tokens.)

The tree that TreeBuilder builds for you is made up of objects of the
class HTML::Element.

If you find that you do not properly understand the documentation
for HTML::TreeBuilder and HTML::Element, it may be because you are
unfamiliar with tree-shaped data structures, or with object-oriented
modules in general. Sean Burke has written some articles for
I<The Perl Journal> (C<www.tpj.com>) that seek to provide that background.
The full text of those articles is contained in this distribution, as:

=over 4

=item L<HTML::Tree::AboutObjects|HTML::Tree::AboutObjects>

"User's View of Object-Oriented Modules" from TPJ17.

=item L<HTML::Tree::AboutTrees|HTML::Tree::AboutTrees>

"Trees" from TPJ18

=item L<HTML::Tree::Scanning|HTML::Tree::Scanning>

"Scanning HTML" from TPJ19

=back

Readers already familiar with object-oriented modules and tree-shaped
data structures should read just the last article.  Readers without
that background should read the first, then the second, and then the
third.

=head1 METHODS

All these methods simply redirect to the corresponding method in
HTML::TreeBuilder.  It's more efficient to use HTML::TreeBuilder
directly, and skip loading HTML::Tree at all.

=method new

Redirects to L<HTML::TreeBuilder/new>.

=method new_from_file

Redirects to L<HTML::TreeBuilder/new_from_file>.

=method new_from_content

Redirects to L<HTML::TreeBuilder/new_from_content>.

=method new_from_string

Redirects to L<HTML::TreeBuilder/new_from_string>.

=method new_from_url

Redirects to L<HTML::TreeBuilder/new_from_url>.

=method new_from_http

Redirects to L<HTML::TreeBuilder/new_from_http>.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc HTML::Tree

    You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/HTML-Tree>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/HTML-Tree>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=HTML-Tree>

=item * Search CPAN

L<http://search.cpan.org/dist/HTML-Tree>

=item * Stack Overflow

L<http://stackoverflow.com/questions/tagged/html-tree>

If you have a question about how to use HTML-Tree, Stack Overflow is
the place to ask it.  Make sure you tag it both C<perl> and C<html-tree>.

=back

=head1 SEE ALSO

L<HTML::TreeBuilder>, L<HTML::Element>, L<HTML::Tagset>,
L<HTML::Parser>, L<HTML::DOMbo>

The book I<Perl & LWP> by Sean M. Burke published by
O'Reilly and Associates, 2002.  ISBN: 0-596-00178-9

It has several chapters to do with HTML processing in general,
and HTML-Tree specifically.  There's more info at:

    http://www.oreilly.com/catalog/perllwp/

    http://www.amazon.com/exec/obidos/ASIN/0596001789

=head1 SOURCE REPOSITORY

HTML-Tree is now maintained using Git.  The main public repository is
L<< {{ $meta{resources}{repository}{web} // die 'no repo' }} >>.

The best way to send a patch is to make a pull request there.

=head1 ACKNOWLEDGEMENTS

Thanks to Gisle Aas, Sean Burke and Andy Lester for their original work.

Thanks to Chicago Perl Mongers (http://chicago.pm.org) for their
patches submitted to HTML::Tree as part of the Phalanx project
(http://qa.perl.org/phalanx).

Thanks to the following people for additional patches and documentation:
Terrence Brannon, Gordon Lack, Chris Madsen and Ricardo Signes.

=cut

